src/mono/netcore/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Mono.cs

```diff
@@ -354,6 +356,86 @@ public static unsafe IntPtr StringToBSTR(string? s)
                 return BufferToBSTR(fixed_s, s.Length);
         }
 
+        internal class MarshalerInstanceKeyComparer : IEqualityComparer<(Type, string)>
+        {
+            public bool Equals((Type, string) lhs, (Type, string) rhs)
+            {
+                return lhs.CompareTo(rhs) == 0;
+            }
+
+            public int GetHashCode((Type, string) key)
+            {
+                return key.GetHashCode();
+            }
+        }
+
+        internal static Dictionary<(Type, string), ICustomMarshaler>? MarshalerInstanceCache;
+        internal static readonly object MarshalerInstanceCacheLock = new object();
+
+        internal static ICustomMarshaler? GetCustomMarshalerInstance(Type type, string cookie)
+        {
+            var key = (type, cookie);
+
+            LazyInitializer.EnsureInitialized(
+                ref MarshalerInstanceCache,
+                () => new Dictionary<(Type, string), ICustomMarshaler>(new MarshalerInstanceKeyComparer())
+            );
+
+            ICustomMarshaler? result;
+            bool gotExistingInstance;
+            lock (MarshalerInstanceCacheLock)
+                gotExistingInstance = MarshalerInstanceCache.TryGetValue(key, out result);
+
+            if (!gotExistingInstance)
+            {
+                RuntimeMethodInfo? getInstanceMethod;
```

Comment:
@jkotas is there coreclr implementation we could cheaply use instead?

Reply:
I do not think so. All interop generation is C++ in CoreCLR.